internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,206 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const lokiParserSize = 4 * 1024
+
+type lokiForwarder struct {
+	serviceName string
+	targetName  string
+	target      plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	stopCh   chan struct{}
+
+	buffer *lokiBuffer
+	client *lokiClient
+}
+
+func newLokiForwarder(
+	serviceName, targetName string, target plan.LogTarget, iterator servicelog.Iterator,
+) (*lokiForwarder, error) {
+
+	labels := map[string]string{
+		"service": serviceName,
+		"manager": "pebble",
+	}
+
+	return &lokiForwarder{
+		serviceName: serviceName,
+		targetName:  targetName,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, lokiParserSize),
+		buffer:      newLokiBuffer(labels, 10, 1*time.Second),
+		client:      newLokiClient(target.Location),
+	}, nil
+}
+
+func (f *lokiForwarder) Forward() {
+	for {
+		// Check if Stop has been called
+		select {
+		case <-f.stopCh:
+			f.flushBuffer()
+			return
+		default:
+			// continue
+		}
+
+		if f.buffer.ReadyToFlush() {
+			f.flushBuffer()
+		}
+
+		if f.parser.Next() {
+			f.buffer.WriteEntry(f.parser.Entry())
+		} else {
+			if err := f.parser.Err(); err != nil {
+				logger.Noticef(
+					"Forwarding logs to target %q: cannot parse log for service %q: %v",
+					f.targetName, f.serviceName, err)
+			}
+		}
+	}
+}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot parse log for service %q: %v",
+			f.targetName, f.serviceName, err)
+		return
+	}
+
+	f.buffer.Empty()
+}
+
+type lokiBuffer struct {
+	labels      map[string]string
+	messages    []string
+	maxMessages int
+
+	deadline *time.Time
+	delay    time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	return &lokiBuffer{
+		labels:      labels,
+		messages:    make([]string, 0, maxMessages),
+		maxMessages: maxMessages,
+		deadline:    nil,
+		delay:       delay,
+	}
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    [ "<unix epoch in nanoseconds>", "<log line>" ]
+	timestamp := fmt.Sprint(entry.Time.UnixNano())
+	message := strings.Trim(entry.Message, "\n")
+	encodedMsg := fmt.Sprintf("[%q,%q]", timestamp, message)
+	b.messages = append(b.messages, encodedMsg)
```

Comment:
Two things:

1) This is actually JSON, and there may be subtle differences between Go's `%q` representation (which is Go syntax) and JSON string representation (which is JSON/JavaScript). In particular, Go supports `\Unnnnnnnn` Unicode escapes, whereas JSON only supports `\unnnn` ones (and would encode things above 0xFFFF as a two-character unicode combination). I realize Unicode smileys in logs is unlikely, but probably will happen in these days of Emoji everywhere.
2) This is probably quite inefficient (and done for every log line): use of `fmt.Sprint` when we know what type it is, string allocation, use of `fmt.Sprintf`, and so on.

We should probably start by fixing the correctness thing, by using `encoding/json` for at least encoding the message. That will probably slow things down a bunch, but could optimize later if needed. Correctness is first. Hopefully we can avoid the `fmt.Sprint` and `fmt.Sprintf` and that will make up for the use of json.

For a bit more efficiency, could use a `json.Encoder` attached to a `bytes.Buffer` which we reset on flush.